packages/attribute-slicer-powerbi/src/AttributeSlicerVisual.ts

```diff
@@ -348,36 +362,90 @@ export default class AttributeSlicer implements powerbi.extensibility.visual.IVi
         }
     }
 
-    /* tslint:disable */
-    /**
-     * The debounced version of the selection changed
-     */
-    private _onSelectionChangedDebounced = _.debounce( /* tslint:enable */
-        (selectedItems: ListItem[]) => {
-            log("onSelectionChanged");
-            const selection = selectedItems.map(n => n.match).join(", ");
-            const text = selection && selection.length ? `Select ${selection}` : "Clear Selection";
-            this.state.selectedItems = selectedItems;
-            this.writeStateToPBI(text);
-        },
-    100);
-
     /**
      * Listener for when the selection changes
      */
-    private onSelectionChanged(newItems: ListItem[]) {
+    private onSelectionChanged(selectedItems: ListItem[]) {
         if (!this.isHandlingUpdate) {
-            this._onSelectionChangedDebounced(newItems);
+            log("onSelectionChanged");
+            const newIds = (selectedItems || []).map(n => n.id).sort();
+            const oldIds = (this.state.selectedItems || []).map(n => n.id).sort();
+            let hasChanges = newIds.length !== oldIds.length;
+            if (!hasChanges) {
+                hasChanges = newIds.some((ni, i) => newIds[i] !== oldIds[i]);
+            }
+            if (hasChanges) {
+                this.state.selectedItems = selectedItems;
+                this.propertyPersister.persist(false, {
+                    merge: [{
+                        objectName: "general",
+                        selector: undefined,
+                        properties: {
+                            selection: JSON.stringify(selectedItems || []),
+                        },
+                    }],
+                });
+
+                const state = this.state;
+                const selItems = state.selectedItems || [];
+                const categories: powerbi.DataViewCategoricalColumn = this.dataView.categorical.categories[0];
+                const source = categories.source;
+                const target: models.IFilterColumnTarget = {
+                    table: source.queryName.substr(0, source.queryName.indexOf(".")),
+                    column: source.displayName,
+                };
+                const filter = new models.BasicFilter(
+                    target,
+                    "In",
+                    selItems.map(n => source.type && source.type.numeric ? parseFloat(n.match) : n.match)
+                );
+
+                const action = selItems.length > 0 ? powerbi.FilterAction.merge : powerbi.FilterAction.remove;
+
+                let applied = false;
+                const applyFilter = () => {
+                    if (!applied) {
+                        applied = true;
+                        this.host.applyJsonFilter(filter, "general", "filter", action);
+                    }
+                };
+
+                this.afterNextUpdate.then(applyFilter);
+                setTimeout(applyFilter, 300);
```

Comment:
Can you give a quick explanation of the logic here, why set a timeout and have all the logic to apply the filter on the next update if it hasn't been applied already instead of just applying it in this function?